File: backend/agents/gridAgent.js
const { callGroq, parseGroqJSON } = require('../utils/groqClient');

/**
 * GRID LOAD & RENEWABLE ENERGY OPTIMIZATION AGENT
 * Analyzes local electrical grid carbon intensity, V2G feasibility, solar availability, and off-peak tariffs.
 */
async function gridAgent(telemetry, chargingResult = null) {
  const currentHour = new Date().getHours();
  const isOffPeak = currentHour >= 22 || currentHour <= 6;
  const isSolarPeak = currentHour >= 10 && currentHour <= 16;

  const fallback = {
    grid_status: isSolarPeak ? 'High Solar Generation (Clean Grid)' : isOffPeak ? 'Off-Peak Tariff Window' : 'Peak Grid Demand',
    carbon_intensity_g_kwh: isSolarPeak ? 180 : isOffPeak ? 240 : 410,
    v2g_capable: true,
    v2g_recommendation: telemetry?.soc > 70 && !isOffPeak ? 'V2G discharge candidate: Earn revenue by supplying power to local microgrid.' : 'Hold charge for vehicle usage.',
    optimal_charging_window: '22:00 - 06:00 (Lowest Carbon & Cost)',
    grid_confidence: 0.91,
  };

  const systemPrompt = `You are the Grid Load & Renewable Energy Optimization Agent.
Analyze electrical grid carbon intensity and V2G (Vehicle-to-Grid) feasibility for EV SOC=${telemetry?.soc}%.
Return JSON:
{
  "grid_status": string,
  "carbon_intensity_g_kwh": number,
  "v2g_capable": boolean,
  "v2g_recommendation": string,
  "optimal_charging_window": string,
  "grid_confidence": number (0-1)
}`;

  try {
    const content = await callGroq([
      { role: 'system', content: systemPrompt },
      { role: 'user', content: `EV SOC: ${telemetry?.soc}%, hour: ${currentHour}` },
    ], 'llama-3.3-70b-versatile', { json_mode: true, max_tokens: 300 });

    return { ...fallback, ...parseGroqJSON(content, fallback) };
  } catch (err) {
    console.warn('[GridAgent] Groq call failed, using rule-based fallback:', err.message);
    return fallback;
  }
}

module.exports = { gridAgent };

File: backend/agents/grid_agent.py
import datetime
from typing import Dict, Any, Optional
from utils.groq_client import call_groq, parse_groq_json

async def grid_agent(telemetry: Dict[str, Any], charging_result: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """
    GRID LOAD & RENEWABLE ENERGY OPTIMIZATION AGENT
    Analyzes local electrical grid carbon intensity, V2G feasibility, solar availability, and off-peak tariffs.
    """
    current_hour = datetime.datetime.now().hour
    is_off_peak = current_hour >= 22 or current_hour <= 6
    is_solar_peak = 10 <= current_hour <= 16

    soc = telemetry.get("soc", 80)

    fallback = {
        "grid_status": "High Solar Generation (Clean Grid)" if is_solar_peak else "Off-Peak Tariff Window" if is_off_peak else "Peak Grid Demand",
        "carbon_intensity_g_kwh": 180 if is_solar_peak else 240 if is_off_peak else 410,
        "v2g_capable": True,
        "v2g_recommendation": "V2G discharge candidate: Earn revenue by supplying power to local microgrid." if (soc > 70 and not is_off_peak) else "Hold charge for vehicle usage.",
        "optimal_charging_window": "22:00 - 06:00 (Lowest Carbon & Cost)",
        "grid_confidence": 0.91,
    }

    system_prompt = f"""You are the Grid Load & Renewable Energy Optimization Agent.
Analyze electrical grid carbon intensity and V2G (Vehicle-to-Grid) feasibility for EV SOC={soc}%.
Return JSON:
{{
  "grid_status": string,
  "carbon_intensity_g_kwh": number,
  "v2g_capable": boolean,
  "v2g_recommendation": string,
  "optimal_charging_window": string,
  "grid_confidence": number (0-1)
}}"""

    try:
        content = await call_groq([
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": f"EV SOC: {soc}%, hour: {current_hour}"}
        ], model="llama-3.3-70b-versatile", options={"json_mode": True, "max_tokens": 300})

        return {**fallback, **parse_groq_json(content, fallback)}
    except Exception as err:
        print(f"[GridAgent] Groq call failed, using rule-based fallback: {err}")
        return fallback
